### ISSUE TITLE
Create a system for managing addresses

### ISSUE THREAD

**Issue by jabdownsmash** [spxtr/p3#4]
In [this branch](/sepharoth213/p3/tree/addresses) I've created a possible way to handle addresses, that is, to have a map of addresses to address objects so that parsing can happen like so:

``` python
#addr.py

addresses = [
#                      name               address     mask        bitshift
    at.IntegerAddress("playerOnePercent", "804530E0", 0x00FF0000, 16),
    at.IntegerAddress("playerTwoPercent", "80453F70", 0x00FF0000, 16),
    at.IntegerAddress("globalFrameCounter", "804D7420", 0xFFFFFFFF, 0),
]

for address in addresses:
    names[address.address] = address

#usage:
for address, value in mw:
    if(address in addr.names):
        parsedValue = addr.names[address].get_value(value)
```

There will likely be many addresses to manage in p3 and I believe that this is a sane way, however, that may not make sense depending on the structure of p3. If p3 moves towards being event-driven I feel that this approach is best, since the above implementation would naturally extend to being able to subscribe to events by name. This would also allow for the Melee game state to be managed as a map of names to values.

**Comment by spxtr** [comment]
This is pretty close to what I'm thinking as well, with one salient difference. I'd like to avoid playerOnePercent, playerTwoPercent, ..., playerOnePosition, playerTwoPosition, ..., and so on. There should be a length-4 list of player states.

**Comment by jabdownsmash** [comment]
You caught me in the middle of working on something related, [here](/sepharoth213/p3/blob/develop/p3/addr.py) is what I came up with for dealing with multiple addresses with offsets:

``` python
addresses += MultipleAddress(at.IntegerAddress,("player#Percent", "804530E0", 16) , 4, "E90")
```

is equivalent to

``` python
addresses += [
    at.IntegerAddress("player1Percent", "804530E0", 16),
    at.IntegerAddress("player2Percent", "80453F70", 16),
#   ...and so on
]
```

In my fork I was planning on using a map instead of a GameState object, and using player1Percent etc to identify values. One benefit I foresee is that I was planning on creating an event dispatcher:

``` python
meleeEvents.addListener('player2Percent', lambda x: print(x))
```

Also, not utilizing lists for player and controller specific addresses keeps from having to do anything special when managing the game state:

``` python
gameState = {}
for address, value in mw:
    if(address in addr.names):
        ao = addr.names[address]
        gameState[ao] = ao.get_value(value)

# elsewhere

if gameState["player2Percent"] > 20:
    # ...
```

**Comment by spxtr** [comment]
It would be nice to have something like `game_state.players[1].percent`, but what you have here is probably easier to understand. Seems reasonable.

**Comment by spxtr** [comment]
What do you think about what I just pushed?

**Comment by jabdownsmash** [comment]
A couple small things:
- Why do the mask before the shift if it's only ever going to have 0s on the right side anyway? Seems to me that the mask is more for specifying the size of the int you want, having it come before the shift means that changing the shift involves changing 2 values instead of 1.
- I really like the name 'handler'; much better than the incredibly confusing 'address' name I was using.

On the whole, I think that the loss of modularity between an address manager and the game state object is fine, perhaps even beneficial, within the context of a Melee AI, but I had envisioned (at least this part of) p3 becoming a general-use Melee API. I've already been utilizing my fork of p3 for data analysis, training neural nets, and data visualization.

When addresses parse simply to names and values, the information is incredibly easy to manipulate and save. I think that the stringly nature of that implementation is a _positive_ thing in the end, especially since a less stringly version can always be made on top of that (something like  `add_multiple_accessors` that returns a `StateManager` given a name-value dictionary), whereas going the other direction is a more difficult task (something that returns a name-value dictionary given a `State`?).

The big deal with the game-state-as-a-dictionary-of-name-value-pairs approach is that address handlers wouldn't have to be hard-coded. This is something I have been struggling with the past few days: while working across 3 languages for MeleeAI, p3, and doing some data visualization in JavaScript, I started playing around with different formats for saving match data. Initially I took the most naive approach possible, simply storing every name-value pair as p3 gets data from Dolphin. I ran into one huge stumbling block with that approach though: I had to manage the _names_ across every platform. If I ever decided a name was going to be reassigned to a different address, or if I wanted to tweak a name, I would have to do it across all the platforms. Even worse, any data that I had saved before then needed to be converted. The solution I came up with was to move the address definitions out to its own file, and store the _Dolphin output_ instead of storing any parsed data. This way, stored data never becomes outdated, and any Melee data parser that I work with can read from the same address definitions file. (granted, this functionality doesn't exist yet, but is trivial to implement) An added benefit of storing data as Dolphin sends it (address string - byte string pairs) is that 'loading' this data is as simple as swapping out `memory_watcher` with this data!

That turned into a wall of text, but what it comes down to is that I would much prefer a parser that returns name-value pairs because of the flexibility that it provides, which I've already been leveraging . However, if you think that's outside of the scope of p3, that's totally understandable, in which case what you've pushed is a really elegant solution to addresses.

**Comment by spxtr** [comment]
I'm hoping to use this as a general-purpose API as well. I'm thinking there should be a folder under `p3/p3` for each common use case, such as a folder for tracking statistics, a folder for playing a simple flowcharting AI, a folder for a neural net AI, and so on.

With that goal in mind, I want to keep the current stuff as simple as possible. I'm thinking there will likely need to be only one or two more file there, something like `__main__.py` that reads command line args and runs the loop, calling into the subpackages when the frame changes.

Does that sound reasonable? It means it'll be a little trickier to do what you're doing already, but I think it'll keep the code much neater in the long run.

**Comment by spxtr** [comment]
I made the change to shift before masking. Remember that python doesn't have a logical shift, so two's complement stuff can be a bother.

**Comment by jabdownsmash** [comment]
I definitely agree that it should be as simple as it can be, but I don't think that `StateManager` reaches that simplicity. I see the following use cases:
1. Getting Dolphin output directly.
   - I described needing this functionality above to create 'replays' of match data.
   - With the current `MemoryWatcher`, this requires converting bytes _back_ to the string that Dolphin outputs. I propose that `MemoryWatcher` be changed to only return the address and the byte string with the null terminator stripped, and that the conversion to bytes happen within the various handlers.
2. Using arbitrary iterables as a substitute for `MemoryWatcher`
   - Generating fake Melee data or using saved data instead of reading output from Dolphin would use this functionality. The above change to `MemoryWatcher` would greatly simplify this, as the handlers would already be designed to parse byte strings. 
3. Getting name-value pairs from parsed information.
   - The use case here is less obvious, but imagine only wanting to know when Melee has moved on to the next frame. With `StateManager`, it would require either hooking in to an event dispatcher for whatever relevant `FrameCount` attribute, or polling the `FrameCount` in your game state constantly.
4. Having a managed game state.
   - This is what `StateManager` provides, but I believe that it does so by _utilizing_ the above functionality. I think that it runs counter to the goal of simplicity to have to build on top of `StateManager` to get any of the above functionality, since none of the other 3 use cases rely on a game state at all.

With that in mind, I propose the following structure:
1. `MemoryWatcher`
   - Simply connects to dolphin and returns a tuple with an address string and a byte string with the null terminator stripped.
2. `ValueHandler`
   - When initialized, reads _an external file_ that has definitions that look something like
   
   ```
   8065CC14 currentMenu int 20 0xF
   804D6CAD stageID int 16
   ```
   - Parses `MemoryWatcher` output and returns name-value tuples.
   - Giving `("804D6CAD","DEADBEEF")` to a `ValueHandler` intialized with the above file would return `("stageID",57005)`
3. `StateManager`
   - Takes name-value tuples, and uses them to manage a `State` object.

I think that this structure allows for all of the use cases above to have as little overhead in both development and performance as possible, while making it very clear how the API works. As always, however, I could be wrong.

Interesting bit about python not having a logical shift, though I think that would still present itself as an issue even if the mask happened before the shift.

**Comment by spxtr** [comment]
1 and 2 are already possible within p3, but not easily across p3/MeleeAI/cpu-level-11. Because memory watchers are just a standard iterator that returns a pair or None, you can just save what comes out into a list and later input that list into a state handler. I'll make this use case explicit in the code.

I was planning on having a driving loop that keeps the current frame and does memory watcher -> state manager handles each iteration. When the frame changes, it calls whatever AI processing or statistics-gathering functions it has hooked up. This is how cpu-level-11 does it. It's also possible to just call them every time we read anything from dolphin, not just a state change. Then it'd be up to the user to skip unless the frame changed.

My gripe with your proposed structure is that it spreads identifiers like `"stageId"` across two files. We have the definition file with the address, name, type, etc. Then we need another line in the state manager that has the name, wrapper, and default. This just feels like too many moving parts, although I see the reasoning.

I definitely want to move the actual memory locations -> handler map creation out of state manager's constructor, and likely to their own file. I'll probably keep it python, though. I don't want to have to do any more text file processing :P

**Comment by jabdownsmash** [comment]
Re: 1 and 2, bytes aren't easily serializable, so doing it with p3 currently would involve converting the bytes back into a string or storing an int. Saving the string that Dolphin sends makes the most sense to me, so that _other_ frameworks don't have to create 2 separate parsers: one that takes Dolphin output, and one that takes serialized bytes.

My worry is mostly one of future-proofing: as the use cases expand the last thing I want is for recorded data to change formats. Someone will eventually have hardware memory reading, and whenever a tool is created that can utilize Melee data I think it's important for there to be a unified layout across the board. 

Re: having a file for address definitions, I was caught up in having a mirrored js library that could access the same definitions, but as long as the format for saving Dolphin data is consistent, then I see that it wouldn't be as much of an issue.

However, I'm not sure I see the point in trying to create a `StateManager` with such complex accessors if it sacrifices the ability to both hook into specific name-value pairs and have databags that are simple dictionaries. Someone working with p3 is going to have to look at the definition file anyway to see what the accessors are named, and the errors for misspelling them are close enough (key error vs object has no property). If you _didn't_ want to offer a `State` class then there would be no identifier spreading.

Is being able to do `self.state.globalFrameCounter` vs `self.state['globalFrameCounter']` really so valuable that you give up being able to do things like this: 

``` python
acceptedInputs = [
    "player1ActionState",
    # "player1Percent",
    # "player1JumpsUsed",
    # "player1InAir",
    # "player1X",
    # "player1Y",
    # "player1Facing",
    # "player1GroundVelocityX",
    # "player1AirVelocityX",
    # "player1AirVelocityY",
    # "player1AttackVelocityX",
    # "player1AttackVelocityY",
]   

def listener (evName, value):
    if evName in acceptedInputs:
        gameState[acceptedInputs.index(evName)] = value
    elif evName == "globalFrameCounter":
        inputs.append(gameState)
        gameState = [i for i in gameState] #create new list with the same values
#later
nn.fit(inputs)
```

This code was taken from my neural net script, where I could restrict what data the net was learning from by removing it from the `acceptedInputs` list. 

Plus, the way that addresses are defined in `addr.py` in my PR, a `State` can be automatically generated in a way that doesn't duplicate identifiers, and doesn't remove the dictionary state from the picture, by extending what `add_multiple_accessors` does to operate on a new object.

**Comment by spxtr** [comment]
We can pickle the list of tuples, or pack into a proto if necessary. It might be worth including a timestamp as well. I think the important point is that I don't want to support lots of special cases within the API, but rather make the API powerful enough that it's easy to do awesome things with it yet simple enough to understand at a glance.

Feature selection for ML will need a layer of indirection already, since we're going to eventually want to create our own features. Still, it's not too bad to just say

``` python
features = [
    self.state.players[0].x,
    self.state.players[0].y,
    self.state.players[0].action_state,
    abs(self.state.players[0].x) < abs(self.state.players[1].x), # Player 0 is closer to the center than player 1
]
...
nn.fit(features)
```

**Comment by spxtr** [comment]
About the shift vs mask, it's definitely correct to do it in this order (shift then mask). Say we want the first byte of `0xF0000000`. If we mask (`0xFF000000`) then shift (`>> 24`) then we end up with `0xFFFFFFF0`, not what we want! Shift then mask gives us `0xF0`.

**Comment by jabdownsmash** [comment]
Support for Python pickles in other languages is poor, and only Python 3 pickles support serializing bytes. I have been using json to serialize to be able to load them in js, which is why I think that `MemoryWatcher` should return the strings it gets rather than returning bytes.

One interesting problem I've been running into re: timestamping is that Dolphin's MemoryWatcher class polls emulated memory for changes every 2ms, and sends addresses to the socket in the order that they are defined in Locations.txt, which (I think) causes the addresses to occasionally be received out of order. You can see effect in [this image](http://imgur.com/jXMT3ax), a plot of a player's position throughout a match. In the big arc on the right side, there are gaps that are followed or preceded by positions being reported twice, which occurs when the players position and the frame counter are received out of order.

You've sold me on `StateManager`, I've been working through various use cases and, while the setup involves using more components than strictly necessary, it is good to have a consistent API across all use cases.

**Comment by spxtr** [comment]
You're definitely right about Dolphin's MemoryWatcher. The goal there is to be very simple and game-agnostic. It polls game memory _without any synchronization_, so there are no guarantees on the order of  what comes out, or the consistency of anything larger than one word.

**Comment by spxtr** [comment]
I'm not really a fan of pickles. JSON is alright, but I'd much rather pack into a proto, especially if we are going to be saving/sharing lots of data.

**Comment by jabdownsmash** [comment]
Protos look like a very apt solution, I wasn't aware of them. The JSONs were a bit size heavy, with a match's worth of data being around 4mb. 

Won't not being able to synchronize with the in-game FPS be a pretty showstopping issue with a lot of applications moving forward? Before I looked into the MemoryWatcher code I mostly expected it to hook into the addresses and send info when they were changed a la a cheat engine.

**Comment by spxtr** [comment]
It's not beautiful, but we can always wait 3-4 ms for further changes after the frame changes. I could suggest lowering the polling delay to 1 ms on Dolphin's side as well, then we'd only need to wait 2 ms. We can also enforce that the frame memory location is the first entry in Locations.txt.

This is part of why I added the timeout to MemoryWatcher :P. Dolphin is pretty primitive as an API, so I wasn't able to hook in to addresses. Polling was the only feasible way.

**Comment by jabdownsmash** [comment]
Would waiting solve the problem? The issue is: whenever you receive a frame count tick and then a position, how do you know if the position was for the frame after the tick or before? 

While I was sitting in the Dolphin IRC for a bit I caught wind of lioncash possibly implementing related functionality. Maybe Dolphin will get less primitive soon lol.

**Comment by spxtr** [comment]
It'd be awesome if we could be sure of it, but for now we have to make do.

I'm going to talk with the dolphin devs a bit more about how to do this properly, but it's a hard problem. We don't want to slow down the emulator by checking a list of addresses each time any part of memory changes.

**Comment by spxtr** [comment]
Going to close this issue, thanks for the discussion. Might be worth opening another to keep talking about accuracy.
